src/kern/npf_ctl.c

```diff
@@ -732,6 +758,70 @@ npfctl_save(npf_t *npf, u_long cmd, void *data)
 	return error;
 }
 
+/*
+ * npfctl_table_replace_nvlist: atomically replace a table's contents
+ * with the passed table data.
+ */
+static int __noinline
+npfctl_table_replace_nvlist(npf_t *npf, nvlist_t *npf_dict, nvlist_t *errdict)
+{
+	npf_tableset_t *tblset;
+	int error = 0;
+
+	npf_table_t *oldt, *newt;
+
+	npf_config_enter(npf);
+	tblset = npf_config_tableset(npf);
+
+	/* Get the entries or binary data. */
+	error = npf_mk_table(npf, npf_dict, errdict, tblset, &newt, 1);
+	if (error) {
+		goto err_config;
+	}
+
+	oldt = npf_tableset_swap(tblset, newt);
+	if (oldt == NULL) {
+		error = EINVAL;
+		goto err_config;
+	}
+	npf_config_sync(npf);
+	npf_config_exit(npf);
+
+	/* At this point, it is safe to destroy the old table. */
+	npf_table_destroy(oldt);
+
+	/* successfully completed */
+	goto out;
+
+err_config:
+	npf_config_exit(npf);
+	npf_table_destroy(newt);
+
+	/* fallthrough */
+out:
+	return error;
+}
+
+int
+npfctl_table_replace(npf_t *npf, u_long cmd, void *data)
+{
+	nvlist_t *request, *response;
+	int error;
+
+	/*
+	 * Retrieve the configuration and check the version.
+	 * Construct a response with error reporting.
+	 */
+	error = npf_nvlist_copyin(npf, data, &request);
```

Comment:
What would free the `request` nvlist? It seems to me that there is a memory leak.

Reply:
Whoops, sorry that's my bad - I modelled this function on npfctl_load(), which doesn't free its request, but I didn't notice that it gets nvlist_destroy()'ed in the called npfctl_load_nvlist() function :(.
I've fixed locally and will push once I've got through the other requested changes.